### ISSUE TITLE
Feature Request: Cloud Function should support user.sendEmailVerification() like client sdk

### ISSUE THREAD

**Issue by inlined** [firebase/firebase-admin-node#46]
[Refiling for user @sun3 from https://github.com/firebase/functions-samples/issues/181]

Feature Request:

The ```user.sendEmailVerification()``` needs to be supported by Firebase Cloud Functions. This is available in client side but not server side. When using Cloud Functions to create new users we also need to automatically send Email Verification before new user can use our apps. This is currently as stopping block for our apps.

At this point the ```user.sendEmailVerification()``` gives error that function is not found.

```
    admin.auth().createUser({
        email: emailAddress,
        emailVerified: false,
        password: password,
        displayName: '', //name,
        disabled: false
    })
        .then(function (user) {
            // A error representation of the newly created user is returned
            console.log("Created Firebase User successfully with id: ", user.uid);
            console.log("user.emailVerified:", user.emailVerified);

            // Send Email Verification
            user.sendEmailVerification()
                .then(function (emailSent) {
                    console.log('emailSent ', emailSent);
                })
                .catch(function (error) {
                    console.log('emailSent error ', error);
                });

    // ... Additional code below
```

Thank you and I am open to any suggestions.

**Comment by Sun3** [comment]
The issue/new feature request is that the ```user.sendEmailVerification()``` function is not available on the firebase admin (firebase cloud functions), but it is available in the client sdk like the one used in angular 4.

From the Documentation when you create a New user. I added the sendEmailVerification() which this function is not available. But it's a must to allow at this time for the new user to verify their email before we allow access to our apps.
```
admin.auth().createUser({
  email: "user@example.com",
  emailVerified: false,
  password: "secretPassword",
  displayName: "John Doe",
  photoURL: "http://www.example.com/12345678/photo.png",
  disabled: false
})
  .then(function(userRecord) {
    // See the UserRecord reference doc for the contents of userRecord.
    console.log("Successfully created new user:", userRecord.uid);
    
    // ****************************************************************
    // Cannot request/call the sendEmailVerification() function, not available
    // Send Email Verification
            user.sendEmailVerification()
                .then(function (emailSent) {
                    console.log('emailSent ', emailSent);
                })
                .catch(function (error) {
                    console.log('emailSent error ', error);
                });
    // End send email verification not available
    // ****************************************************************


  })
  .catch(function(error) {
    console.log("Error creating new user:", error);
  });
```

Please let me know if this explains the issue.
Thanks.

**Comment by rkpradeep20** [comment]
Yes, would like this feature in cloud functions. user.sendEmailVerification()

**Comment by coreybutler** [comment]
The [docs](https://firebase.google.com/docs/auth/extend-with-functions?authuser=0) make it seem like its possible to verify a new user's email address with the firebase email validation using a cloud function. Lack of this feature _forces_ developers to use an external service (which requires an upgraded paid plan). Requiring a third party dependency for something Firebase already does well is clunky for developers and confusing.

**Comment by bojeil-google** [comment]
Hey @coreybutler, the email verification Firebase Auth sends is not a welcome nor a goodbye email (the link you pointed it out). Nor should it be used for that.

You also have the option to send an email verification client side for free. So if you want to send an email verification, you can always do it for free.

Let's keep these issues separate (sending email verifications, vs sending other personalized emails for non-auth specific reasons). We acknowledge the value of sending an email verification, server side via the admin SDK, and we are looking into it.

**Comment by coreybutler** [comment]
@bojeil-google - thanks.

Purely for context, my "app" doesn't have a front end. In my case, I'm using a cloud function to serve as a Docker Authentication Proxy... 100% server-side. Users are added directly through the Firebase console.

If this is a separate issue (which is fine), the docs should clarify. Perhaps indicate what limitations exist. A document titled "Extend Firebase Authentication with Cloud Functions" and a subtitle of "Trigger a function on user creation" seems like it would encompass email verification.

**Comment by bojeil-google** [comment]
Hey @coreybutler, to help unblock you, you can always use the client SDK to send the email verification from a Firebase Function. You'd need to require firebase module.
On user creation, you do the following:
1. get the `uid` of the user created from the onCreate trigger.
2. Using admin SDK, mint custom token via `createCustomToken` for that user.
3. Using client SDK, `signInWithCustomToken` using that custom token.
4. Send email verification via `currentUser.sendEmailVerification`.

**Comment by PoiScript** [comment]
+1, for now, I'm sending verification email by hitting [the restful api](https://firebase.google.com/docs/reference/rest/auth/#section-send-email-verification). What a mess..

**Comment by inlined** [comment]
BTW, I'm trying to catch examples of sub-optimal Promise usage in the wild and wanted to point out the one in this thread, CC @Sun3 Promise code should almost never repeatedly nest. These aren't callbacks anymore:

```javascript
    admin.auth().createUser({
        email: emailAddress,
        emailVerified: false,
        password: password,
        displayName: '', //name,
        disabled: false
    }).then(function (user) {
        // A error representation of the newly created user is returned
        console.log("Created Firebase User successfully with id: ", user.uid);
        console.log("user.emailVerified:", user.emailVerified);

        // Send Email Verification
       return user.sendEmailVerification();
    }).then(function (emailSent) {
       console.log('emailSent ', emailSent);

    }).catch(function (error) {
       console.log('emailSent error ', error);
    });
    // ... Additional code below
```

This code should hopefully be more readable. It also fixes two bugs: 

1. `user.sendEmailVerificaiton` wasn't being returned, so the outer promise was resolving early to `undefined`
2. Because the `catch` block was on the outer promise, there was no handler for a failure to create the new user (may have been intentional from the error message, but you should always handle rejection.

**Comment by sweetcoco** [comment]
@bojeil-google While that would work to unblock developers who need this urgently, its an incredibly elaborate workaround once you figure in the need for service account certs for minting custom tokens... in my case where i have multiple environments where i would need to manage certs based on environment, its just a lot mental overhead. It would be nice if cloud functions didn't need service account certs anyway.

Any word if this feature is being worked on?

**Comment by bojeil-google** [comment]
The feature is on our list and we acknowledge its importance. We just have a lot of feature requests and many are currently critical to developers and not possible client side or admin side. I would prioritize those higher.

If you are having a hard time minting custom tokens, you can just send the ID token to an HTTP endpoint you host in Firebase Functions and use the REST API to send the email verifications:
https://firebase.google.com/docs/reference/rest/auth/#section-send-email-verification

**Comment by sweetcoco** [comment]
@bojeil-google Thanks a ton!

**Comment by coreybutler** [comment]
@bojeil-google - sorry for the delayed response. Are you suggesting the firebase client SDK can be used from Node (i.e. within the function)? Just to reiterate, I have no front end and no browser, so all logic must be encapsulated within the Firebase function (which is what I was trying to do). Step 3 is what I'm questioning the viability of. I thought the client SDK relied on browser-specific capabilities, such as XHR, which would make this impossible. I'd love to be wrong about that ;-)

The goal of my project is for users to login to Docker via a terminal/shell, using Firebase auth. The Docker agent handles all of the handshaking, but uses HTTP requests to authenticate users. Basically, Docker attempts to login via basic auth first, so my function extracts the credentials and matches them up with a Firebase user. That was pretty simple. 

I wanted my code to assure the user had a verified email address. We know and trust users, but we don't always trust someone to type in the right email... typos are rampant. So, the goal was to setup a function that would send the verification email whenever a new user was created, regardless of how they're created (which is exclusively through the web console at the moment).

I know this isn't a traditional use of Firebase, and our goal is to eventually have a more traditional UI/self-registration portal, which would make all of this moot. However; we don't have enough time for a project of that scale right now. Personally, I'd be even happier if there were an option in the Firebase console to automatically send new users a verification email so we don't have to write this functionality ourselves... but I'm also perfectly content doing it myself if it's possible.

**Comment by inlined** [comment]
Yes, the client SDK works in Node. We actually use the Client SDK in Node as part of `firebase-tools`, so this is pretty unlikely to change any time soon.

**Comment by rwmb** [comment]
I have the latest version of the admin sdk but I don't see this feature implemented. Do we still have to login as the user in order to send the verification?
If so, I couldn't find a solution or reason for this issue to be closed.
Thanks a lot.

**Comment by hiranya911** [comment]
This issue is not closed.

**Comment by rwmb** [comment]
Oh ok, my bad. I saw the reference and mistaken it for an action in this issue.
Thank you @hiranya911
So +1 for this feature 👍

**Comment by lingxiao** [comment]
Any updates on this issue? It seems like a pretty crucial feature...

**Comment by fcarreno** [comment]
Agree this will be a great feature to add to the Admin SDK.

I understand an ID token would really not be required to perform these type of operation from the server side and assume the plan is to just use service account credentials, just like with other sever side/admin APIs.

That said, with the current situation, I gave it a shot with a combination of sign up + send email verification REST APIs: 

https://firebase.google.com/docs/reference/rest/auth/#section-create-email-password 
(returns an ID token if successful) 

https://firebase.google.com/docs/reference/rest/auth/#section-send-email-verification

And that worked well.

Did not check options available to customize the email using this approach, neither any possible throttling issues involved, given this will be triggered from the same IP address, though....

**Comment by Rekodr** [comment]
Is this feature already available ? ; this is really needed.

**Comment by j-low** [comment]
Formally seconding (thirding? fourthing?) this feature.  Also, a bit more clarity in the docs that you  _cannot_ issue a verification email from Cloud Functions would be helpful.

Currently the Auth API for the client and Cloud Functions _appear_ very similar (e.g. `firebase.auth().currentUser...` vs `functions.auth.user()...`) and it took a lot of digging and testing to figure out that I could call `sendEmailVerification()` from the former but not the latter.  The flow has the appearance of convenience but it really felt like a head-fake when I learned that sending an email verification in a logical create-user flow (i.e. within the `onCreate` callback) could not actually be handled nearly as simply as the client-side feature seemed to imply.

**Comment by kkukshtel** [comment]
Fithing (?) this as well. Would love to have this ability be a single call vs. needing to leverage the REST API.

**Comment by kkukshtel** [comment]
Okay, posting this so others have a reference implementation but also to really show the need for having this wrapped up in a simple function. A crazy thing I realized too was the difference between Auth customTokens and IdTokens, and that the Admin Auth SDK is only able to generate the former. This means you have to not only need a REST SDK call to send the email, but also need to add in an additional call to exchange your customToken for an idToken. The Admin Auth SDK also has no method for verifying a customToken to generate an idToken (which is what it looks like the REST SDK does).

This uses the Admin SDK + REST SDK (and the request.js npm module) to send a verification email from the server when a new user signs up and all you have at that point is their UID:

```js
auth.createCustomToken(userRecord.uid).then(function(customToken) {
    var idTokenRequestOps = 
    {
        url:'https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyCustomToken?key=[APIKEY]',
        json: true,
        body: { "token" : customToken, "returnSecureToken" : true}
    };
    request.post(idTokenRequestOps, function optionalCallback(err, httpResponse, body) {
        if (err) {
            return console.error('unable to post request to send swap custom token for id token', err);
        }
        else {
            if(body.error) {
                console.log("unable to swap custom token for idtoken with error", body.error);
            }
            else {
                var emailVerificationOps = 
                {
                    url:'https://www.googleapis.com/identitytoolkit/v3/relyingparty/getOobConfirmationCode?key=[APIKEY]',
                    json: true,
                    body: { "requestType" : "VERIFY_EMAIL", "idToken" : body.idToken}
                };
                request.post(emailVerificationOps, function optionalCallback(err, httpResponse, body) {
                    if (err) {
                        return console.error('unable to post request to send verification email', err);
                    }
                    else {
                        if(body.error) {
                            console.log("unable to send email verification with error",body.error);
                        }
                        else {
                            console.log("sent verification email! server responded with",body);
                        }
                    }
                });
            }
        }
    });
});
```

**Comment by justdan0227** [comment]
So I'm going to ask. Has this gotten any better.  Just a simple create a user and send the email verification link.

**Comment by ananthu-confianz** [comment]
Thank you @kkukshtel Its works like a charm!
But in order to work with createCustomToken() function we have to nitialize the admin with a service account like follow.
Refer: https://stackoverflow.com/questions/42717540/firebase-cloud-functions-createcustomtoken

```const admin = require('firebase-admin');
var serviceAccount = require('./serviceAccountKey.json');
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: "https://< Project ID>.firebaseio.com"
});```

Here is my complete code:

```const cloudStorageConfig = require('./cloud-storage-config.json');
const admin = require('firebase-admin');
var serviceAccount = require('./serviceAccountKey.json');
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: "https://catalog-e8e80.firebaseio.com"
});
const request = require('request');

exports.customerSignUp = functions.https.onRequest((req, res) => {
  cors(req, res, () => { });
admin.auth().createUser({
      email: req.body.email,
      password: req.body.password
    }).then(function (userRecord) {
      sendVerificationEmail(userRecord).then(()=>{
res.status(200).send({ "success": true, uid: userRecord.uid });      
).catch((error)=>{
res.status(200).send({ "success": false, error: error });
})
    })
});

function sendVerificationEmail(userRecord) {
  return admin.auth().createCustomToken(userRecord.uid).then(function (customToken) {
    var header = {
      'Content-Type': 'application/json'
    }
    return request({
      url: "https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyCustomToken?key=" + cloudStorageConfig.apiKey,
      method: 'POST',
      json: { "token": customToken, "returnSecureToken": true },
      headers: header
    },
      function (error, response, body) {
        if (error) {
          console.log('unable to post request to send swap custom token for id token', error);
          return;
        }
        if (body.error) {
          console.log('unable to swap custom token for idtoken with error', body.error);
          return;
        } else {
          return request({
            url: "https://www.googleapis.com/identitytoolkit/v3/relyingparty/getOobConfirmationCode?key=" + cloudStorageConfig.apiKey,
            method: 'POST',
            json: { "requestType": "VERIFY_EMAIL", "idToken": body.idToken },
            headers: header
          },
            function (err, response, body) {
              if (err) {
                console.error('unable to post request to send verification email', err);
              } else {
                if (body.error) {
                  console.log("unable to send email verification with error", body.error);
                } else {
                  console.log("sent verification email! server responded with", body);
                }
              }
              return;
            }
          );
        }
      }
    );
  });
}```

**Comment by yuliankarapetkov** [comment]
Hey, any updates on this?

**Comment by justdan0227** [comment]
I'm having to call a nodejs module with a rest api in order to do it unless this has been now included

**Comment by rhass99** [comment]
> Hey @coreybutler, to help unblock you, you can always use the client SDK to send the email verification from a Firebase Function. You'd need to require firebase module.
> On user creation, you do the following:
> 
> 1. get the `uid` of the user created from the onCreate trigger.
> 2. Using admin SDK, mint custom token via `createCustomToken` for that user.
> 3. Using client SDK, `signInWithCustomToken` using that custom token.
> 4. Send email verification via `currentUser.sendEmailVerification`.

How would I initialize the firebase app on the cloud functions?
firebase.initializeApp({config}) doesnt work, also firebase.initializeApp() doesnt work

**Comment by LoneKlingon** [comment]
I'd also like to request this feature at some point as well. While it probably isn't the most important feature, the alternative presented, creating a custom token and then posting to googleapis, isn't as simple as it could be. I am having issues with the post request accepting my custom id token and as far as I can tell I followed all of the documentation/steps. I can think of a client side work around but it would be easier to do this all server side.

**Comment by rhass99** [comment]
### This is how I implemented it successfully following @bojeil-google earlier comment
#### Firebase Cloud function (background) triggered with every new user created
- This function sends a "user" object to your api endpoint

```javascript
const functions = require('firebase-functions');
const fetch = require('node-fetch');

// Send email verification through express server
exports.sendVerificationEmail = functions.auth.user().onCreate((user) => {
// Example of API ENPOINT URL 'https://mybackendapi.com/api/verifyemail/'
  return fetch(<API ENDPOINT URL>, {
    method: 'POST',
    body: JSON.stringify({ user: user} ),
    headers: {
      "Content-Type": "application/json"
    }
  }).then(res => console.log(res))
    .catch(err => console.log(err));
});
```

#### Server Middleware code
- verifyEmail here is used as middleware
```javascript
// File name 'middleware.js'
export const verifyEmail = async (req, res, next) => {
  const sentUser = req.body.user;
  try {
    const customToken = await admin.auth().createCustomToken(sentUser.uid);
    await firebase.auth().signInWithCustomToken(customToken);
    const mycurrentUser = firebase.auth().currentUser;
    await mycurrentUser.sendEmailVerification();
    res.locals.data = mycurrentUser;
    next();
  } catch (err) {
    next(err);
  }
};
```

#### Server code
```javascript
// Filename 'app.js'
import express from 'express';
import bodyParser from 'body-parser';
// If you don't use cors, the api will reject request if u call it from Cloud functions
import cors from 'cors';
import { verifyEmail } from './middleware'

app.use(cors());
app.use(bodyParser.urlencoded({
    extended: true,
}));
app.use(bodyParser.json());

const app = express();
// If you use the above example for endpoint then here will be
// '/api/verifyemail/'
app.post('<PATH TO ENDPOINT>', verifyEmail, (req, res, next) => {
    res.json({
    status: 'success',
    data: res.locals.data
  });
next()
})
```

This endpoint will return back the full user object and will send the verification email to user.

I hope this helps.

**Comment by IngAjVillalon** [comment]
Apparently, there are many of us who are asking firebase developers to add this feature to Node.js firebase-admin SDK.
**Please!**

**Comment by thereis** [comment]
Any update?

**Comment by bojeil-google** [comment]
The admin SDK recently started supporting [generation of email links](https://firebase.google.com/docs/auth/admin/email-action-links) for email verification, password reset and email link sign-in. This allows you to fully build and customize your own email templates. However, you will need to use your own or a third party SMTP sever to send the email.

**Comment by kkukshtel** [comment]
@bojeil-google this is great! Is this considered "done" from the FB team? I'd personally love to not have to rely on a third party service for this, it should just be built into Firebase or have some setting that lets you toggle it.

**Comment by fatihkurnaz** [comment]
Create user and send verification email. I think that these 2 features are almost inseparable. While you may consider merging under single cloud function, it was disappointed  that we can not even use them separately.

**Comment by naclcaleb** [comment]
The REST API allows sending an email through Firebase via this endpoint: https://firebase.google.com/docs/reference/rest/auth/#section-send-email-verification

It seems pretty easy to use: just give it an email address and it attempts to send verification to the address.

**Comment by bmcilw1** [comment]
The REST API is not an ideal solution, but an ok workaround. In my opinion, ideally, the admin SDK should have a function to sendEmailVerification() for the current user.

**Comment by Jaimeloeuf** [comment]
Been almost 3 years.....

**Comment by dungahk** [comment]
I do not think this will ever be implemented, they added other functionalities such as https://firebase.google.com/docs/auth/admin/email-action-links that pretty much covers 90% of this and the way I see it, that is the best solution going forward.

Yes, you would still need to use your own email service or a third-party one, but nowadays you can get it for free (SendGrid, for example).

**Comment by mariotacke** [comment]
Just came across this issue :/. Since this is already working for the client SDK, what is the suggested work-around for now without relying on a 3rd party service?

**Comment by coreybutler** [comment]
I'll just go ahead an ask.... is there actually any community interest in a 3rd party service to do this? Since my original posts, I've resolved this a number of different times. It's not rocket science from a code perspective, it's just a PITA to setup every time. 

I was thinking along the lines of configuring a simple `onCreate` function to send a web request off to a service when a user is created.... service does everything else and could conceivably send a webhook back to systems once the verification is complete.

I don't want to hijack this thread, so ping me separately if anyone is interested in something like this.

**Comment by mariotacke** [comment]
I found a work-around that works well enough for my use case, see below. I'm not sure if this is best practice, but I wanted to keep the emails exactly the same between the server and client requests. Would love to hear about any flaws with this implementation 💡 

As suggested above, it uses a three step process to do this:
1. Acquire a `custom token` via the admin sdk's `createCustomToken(uid)`
2. It converts this custom token to an `idToken` via the API
3. It invokes the `send email verification` endpoint on the API

```
const functions = require('firebase-functions');
const fetch = require('node-fetch');
const admin = require('firebase-admin');

const apikey = functions.config().project.apikey;
const exchangeCustomTokenEndpoint = `https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken?key=${apikey}`;
const sendEmailVerificationEndpoint = `https://identitytoolkit.googleapis.com/v1/accounts:sendOobCode?key=${apikey}`;

module.exports = functions.auth.user().onCreate(async (user) => {
  if (!user.emailVerified) {
    try {
      const customToken = await admin.auth().createCustomToken(user.uid);

      const { idToken } = await fetch(exchangeCustomTokenEndpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          token: customToken,
          returnSecureToken: true,
        }),
      }).then((res) => res.json());

      const response = await fetch(sendEmailVerificationEndpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          requestType: 'VERIFY_EMAIL',
          idToken: idToken,
        }),
      }).then((res) => res.json());

      // eslint-disable-next-line no-console
      console.log(`Sent email verification to ${response.email}`);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(error);
    }
  }
});

```

**Comment by dmvvilela** [comment]
It's 2020.. nothing on this yet?? C'mon firebase team! I lost hours trying this because it was intuitive to make it server side and even on client side is not working.

**Comment by quantuminformation** [comment]
I'm pretty stumped why the admin would have less functionality than the insecure front end

**Comment by digimbyte** [comment]
I find the lack of user verification and email/password reset from the admin sdk to be a flawed and lacking feature

**Comment by Ranguna** [comment]
+1 for this.
I'm currently using a combination of calling the identity toolkit login with password + send Oob code endpoints to do this, which just feels like a big workaround.

**Comment by Sun3** [comment]
Firebase team, can you please shed light on this issue? The Admin SDK lack of using the sendEmailVerification() link is a serious issue not being address. It should not rely on third party to actually send the email out, it's a standard Firebase auth fuction.

I first reported this issue back Jun 22, 2017.

Thank you.

**Comment by stehag** [comment]
Not having this feature after so many years makes one wonder what is happening, the developer experience is really bad, of course this should have been implemented at the same time as it was possible to create users with the Admin SDK.

**Comment by inlined** [comment]
The problem here is that the backend API is designed to extract the user info from the current token rather than a parameter to the API. This is an older style of Google APIs. I don't foresee a backend API rewrite getting prioritized any time soon. I'd be curious though whether there's a reasonable workaround by creating a custom auth token with the admin SDK and then calling sendEmailVerification with the client SDK. It's a terrible hack but might be better than nothing.

**Comment by Ranguna** [comment]
@stehag  I'm not sure we should say that the developer experience is "really bad" just because of a few missing features here and there. But yeah, this one is a real must for a lot of things.

**Comment by inlined** [comment]
I've re-escalated this internally. Has anyone who has been waiting all this time tried just using the client SDK from the server side? I would expect that you could create a custom token in the admin SDK, pass it to the client SDK, and then call the appropriate method. I get that this is a kludge, but it's better than waiting on a complete backend API rewrite to support this edge case (sadly that's what would be required).

**Comment by myspivey** [comment]
I can confirm that works as it is what I did for the project I was on. It was very "hacky" and did not look good and only worked for custom auth flows not SSO. This was for a project where admins had to create users as self signup was not allowed. Once they had their custom signon, at that point they could switch to SSO.

**Comment by dungahk** [comment]
A few years ago I did exactly that, I guess that still works but I can confirm it used to work around 3 years ago. (Using the client SDK)

**Comment by Ranguna** [comment]
@inlined 
> Has anyone who has been waiting all this time tried just using the client SDK from the server side? 

Isn't the client sdk rate limited for client use ?

**Comment by inlined** [comment]
Rate limiting shouldn't be a problem AFAICT.

**Comment by Ranguna** [comment]
So you are saying that a public facing lib is not rate limited by IP ?
Doesn't sound very realistic, I'd search the docs but they usually aren't very explicit about these things.

**Comment by inlined** [comment]
I see a ratelimit error in their service definition, but don't know (and probably shouldn't say) what the limit is. Generally even IP ratelimts are set to "reasonable" levels because ISPs in some countries NAT massive numbers of clients to the same IPv4. Either way, the rate limit is guaranteed to be greater than 0, so this is better than waiting for an architecture rewrite.

**Comment by Ranguna** [comment]
@inlined , in that case, this should be a better solution https://github.com/firebase/firebase-admin-node/issues/46#issuecomment-625026299, since it shouldn't be rate limited the same way a public facing lib is.
Unless they are using the same underlying endpoints.

**Comment by crossan007** [comment]
So, I have https://github.com/firebase/firebase-admin-node/issues/46#issuecomment-625026299 working in local emulators when I manually specify the REST API key (`const apikey = functions.config().project.apikey;`).

Is there a way in a production environment to get the REST API key?  or will I need to manually add that to the functions config via  `firebase functions:config:set`?

**Comment by inlined** [comment]
The Firebase Config does not include an API key since the Firebase Config is for backend development and the API key is ostensibly to label a client. You can inject it with `functions:config:set` though.

**Comment by MrVentzi** [comment]
+1, we want to see this!

**Comment by Migaelh** [comment]
I found a method on the [Firebase Auth documentation](https://firebase.google.com/docs/auth/admin/email-action-links) where you can generate a `email verification link`, `email password link` or `email link for sign-in`.

You can also pass a redirect URL (and other info) where you can handle any callbacks once the use clicks on the link. 

You just need to send the email yourself. 

This is what I have done: 

```JavaScript

let displayName = 'John Doe'
let email = 'to@mail.com'
//Generate the email verification link
let emailVerificationLink = await admin.auth().generateEmailVerificationLink(email, { url: `SOME_REDIRECT_URL?with=params` })
let mail = JSON.parse(process.env.FIREBASE_CONFIG).mail

//construct the email 
const mailTransport = nodemailer.createTransport({
    service: 'gmail',
    auth: {
      user: mail.email,
      pass: mail.password,
    },
  })

const mailOptions = {
    from: `"${APP_NAME}" <${mail.email}>`,
    to: email,
    subject: `Email verification for ${APP_NAME}`,
    text: `Hello ${displayName})}

        Please follow this link to verify your email address for the ${APP_NAME}
        ${emailVerificationLink}

        Thanks
        Your ${APP_NAME} team
    `,
    html: `
        <p>Hello ${displayName}</p>
        <p>Please follow this link to verify your email address for the ${APP_NAME}</p>
        <p><a href='${emailVerificationLink}'>Verify Email</a></p>
        <p>Thanks</p>
        <p>Your ${APP_NAME} team</p>
    `
  };

  try {
  //send the email
    await mailTransport.sendMail(mailOptions);
  } catch(error) {
    console.error('send email error', error) 
  }
```
I am using the [NodeMailer](https://nodemailer.com/about/) package for sending the email

**Comment by samasthwafer** [comment]
7 years later and no feature still?

**Comment by MadhavKanna** [comment]
I really do need this, guess I'll just have to implement the workaround. Any updates on when the solution should come out?

**Comment by digimbyte** [comment]
> I really do need this, guess I'll just have to implement the workaround. Any updates on when the solution should come out?

something like this can be done:
(doesn't need the client sdk)
const admin = require('firebase-admin');
const axios = require('axios');

admin.initializeApp({
  credential: admin.credential.applicationDefault(),
});

const generateCustomToken = async (uid) => {
  try {
    const customToken = await admin.auth().createCustomToken(uid);
    console.log('Custom Token:', customToken);
    return customToken;
  } catch (error) {
    console.error('Error creating custom token:', error);
  }
};

const authenticateWithCustomToken = async (customToken) => {
  const url = `https://identitytoolkit.googleapis.com/v1/accounts:signInWithCustomToken?key=${process.env.FIREBASE_API_KEY}`;
  try {
    const response = await axios.post(url, {
      token: customToken,
      returnSecureToken: true,
    });
    console.log('ID Token:', response.data.idToken);
    return response.data.idToken;
  } catch (error) {
    console.error('Error authenticating with custom token:', error);
  }
};

const sendVerificationEmail = async (idToken) => {
  const url = `https://identitytoolkit.googleapis.com/v1/accounts:sendOobCode?key=${process.env.FIREBASE_API_KEY}`;
  try {
    const response = await axios.post(url, {
      requestType: 'VERIFY_EMAIL',
      idToken: idToken,
    });
    console.log('Verification email sent:', response.data);
  } catch (error) {
    console.error('Error sending verification email:', error);
  }
};

// Replace with your user's UID
const uid = 'your-user-uid';

generateCustomToken(uid)
  .then(customToken => authenticateWithCustomToken(customToken))
  .then(idToken => sendVerificationEmail(idToken));

**Comment by emre-boz** [comment]
When you use sendPasswordResetEmail from the frontend, it's possible to determine whether an email address is registered in the Firebase project. This exposes a security vulnerability. Furthermore, if the email does belong to a registered user, a password reset email is actually sent to that address.

Client-side protections can easily be bypassed, which is why the sendPasswordResetEmail function should ideally be available on the backend, where proper validation and rate limiting can be enforced securely.

This is not merely a feature request. It is a critical necessity.
